src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+#include "horizontal_fusion_impl.hpp"
+#include <migraphx/iterator_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+static unsigned opcode_bits  = 16;
+static unsigned hash_id_bits = 16;
+static unsigned filter_bits  = 8;
+static unsigned kernel_bits  = 8;
+
+// Register a single operation.
+// 1st arg: operation name. 2nd arg: encoding function.
+void horizontal_fusion_impl::register_op(const std::string& name, encoder func, int flag)
+{
+    op_registry[name] = std::move(func);
+    op_flag[name]     = flag;
+}
+
+// Register operations.
+void horizontal_fusion_impl::register_all()
+{
+    register_op("gpu::convolution", encode_conv_common, 1);
+    register_op("gpu::conv_bias_relu", encode_conv_common, 1);
+    register_op("hip::add_relu", encode_common, 0);
+    register_op("convolution", encode_conv_common, 1);
+    register_op("add", encode_common, 0);
+    register_op("relu", encode_common, 0);
```

Comment:
These should be stored as a static map.

Reply:
fixed

Reply:
Using static map causes the following tidy error:

/var/jenkins/workspace/hX_AMDMIGraphX_horizontal_fusion/src/opt/horizontal_fusion_impl.cpp:12:66: error: initialization of 'op_registry' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp,-warnings-as-errors]

std::unordered_map<std::string, encoder> horizontal_fusion_impl::op_registry =

/var/jenkins/workspace/hX_AMDMIGraphX_horizontal_fusion/src/opt/horizontal_fusion_impl.hpp:106:53: note: possibly throwing function declared here

    static std::unordered_map<std::string, encoder> create_op_registery();

Reply:
As I mentioned [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/251/files#r286212670), it needs to be a local static variable in a function:

```cpp
const std::unordered_map<std::string, encoder>& get_op_registery()
{
    static std::unordered_map<std::string, encoder> m = {
        {"gpu::convolution", encode_conv_common},
        {"gpu::conv_bias_relu", encode_conv_common},
        {"hip::add_relu", encode_common},
        {"convolution", encode_conv_common},
        {"add", encode_common},
        {"relu", encode_common},
    };
    return m;
}
```

This will ensure it is initialized once but cannot be modified.